### DISCUSSION TITLE
Website iFrame Preview Interface

### DISCUSSION THREAD

**Discussion by benhaynes** [directus/directus#5897] [Feature Requests]
### Summary

Add an inline-preview option on the item detail page that allows the editor to see the output of the content in an iFrame. 

### Basic Example

https://user-images.githubusercontent.com/22577866/231797682-3d877ec4-a2a2-44a5-a3cb-156753a572d2.mp4

### Motivation

When working in a headless environment, it can be difficult to mentally visualize what you're data is going to look when it's actually on the output website. By allowing the end-project website/spa to be rendered inline, the editor can immediately see the effects of their changes.

### Detailed Design

A new setting will be added to the collection configuration that allows the user to configure the URL that will be used to preview the content. This URL also allows for parameters about the current item that's previewed. This includes any of the fields of the collection, and optionally the auth token of the current user. This allows the front-end page to render the current item that's being previewed in a private preview environment.

Once a collection has this URL configured, a new header button will be available in the header bar that toggles the preview mode. The preview (by default) renders on the right-half of the edit page, and can be dynamically resized.

#### Must Have:

- Ability to configure parameterized preview URL in data-model settings
- Inline iframe that renders configured URL

#### Should Have:

- Iframe refresh on save of the parent item
- Ability to render preview in separate browser window
- Ability to resize the preview window
- Button to manually refresh the iframe-d page

#### Could Have:

- Devtools on the iframe preview window, like media queries or inspectors

#### Won't Have:

- Real-time sync between edit form and page

### Drawbacks

It's a new opt-in tool, so shouldn't introduce any new drawbacks to currently existing feature set.

### Alternatives

You could use a button-link interface to deep-link to a preview page, though this doesn't allow you to show the preview side by side with the editing experience.

### Adoption Strategy

Admins can configure any URL to be previewed in-app. In order to maximize the usefulness of this feature, the website could be made aware of an inline-auth token to then render items that would normally not be available for public consumption yet.

### Unresolved Questions

Not really an unresolved question for this MVP iteration of the feature, but an obvious next question here would be: "How can we combine this with the real-time endpoints to allow for live 1-1 editing in the preview?"

**Comment by altezza04** [comment]
that's a very essential feature for CMS, and a lot reason  many people still use wordpress instead of api based cms like directus (at least for our client) is they don't usually include preview or live editor features. 

i also suggest if somehow in the SDK can have a set of functions to added a live editor in the front end coding.

**Comment by philleepflorence** [comment]
**Interesting Feature Request!**
On a couple D8 instances, the web app styles are applied to a read-only view of the current item, even before saving.

> Got the idea from Directus actually, switched to CSS custom properties (variables), so some design-centric admins could tweak the styles within Directus, without touching the CSS Stylesheets...made it easier to load the rules and variables into the web component.

It would be nice to see ideas on implementation for websites and web apps.

Is this to preview the current item content in a responsive container?

**Reply by benhaynes** [reply]
Yes, the thinking is that you could have your Digital Experience (eg: website) loaded into the iFrame Interface, and get a live preview of what your data would look like (some standard for passing draft data to the page). Could have a button to refresh and options for auto-reload on data change.

**Reply by philleepflorence** [reply]
I would be happy to implement this feature for D9...lemme list how it is currently done to get feedback and ideas on usability:

### Context

A couple implementations didn't favor iFrame for some clients, we had control over Directus and Web Apps, modern browsers were recommended.

### D8 Implementation

- Web Apps are mostly Vue or React, so there was some room to only pull relevant layouts (e.g. a blog post in a plain layout with no navigation or header).
- The "site" loads inside an HTML Web Component and Shadow DOM, draft data from Directus is passed through or events dispatched back, and the apps respond more to sizing.
- Content contributors can actually preview compiled content, even though they cannot publish without moderators, all within the Data Manager.

### Caveat

- My thinking is the Directus SDK would be an important Bridge to prevent complicating the access to the DB draft items for the roundtrip API journey, especially if the "site" has to be virtually isolated and sandboxed.
- Has anyone encountered issues with iFrames, Cross Domain Communication, and HTTP Headers?

**Reply by benhaynes** [reply]
I think the best way to support this is to:

1. Directus provides a "Preview" Interface that renders a webpage in an iFrame using this URL option:
    * URL: `https://example.com/articles/{{slug}}`
2. The URL could also attach a query string with a static token for auth. This could be the current user's, or an Interface option for overriding. This allows elevated privileges for accessing draft content.
3. The website/webapp would need to have some additional logic added to support the query string token... allowing rendering the draft content preview.

@rijkvanzanten thoughts?

**Reply by oldskool73** [reply]
Hey all, new to your project but am interested in this type of feature being added. Might I suggest, if you haven't already, you take a look at the way Storyblok implements this (https://www.storyblok.com/docs/guide/essentials/visual-editor), IMHO it's pretty elegant and simple, you just include a small snippet they provide into your page and it gives you one or two way real time updates with very little setup required.

**Comment by madc** [comment]
Craft CMS also comes with this feature and it seems really well implemented. I could not find it in the docs, but here is a quick post about it: [ Headless Preview in Craft CMS](https://nystudio107.com/blog/headless-preview-in-craft-cms). It's been in there since version [3.2](https://craftcms.com/blog/craft-32).

Is this something that could land in the core in the near(er) future, or would it feasible to build this as a plugin?

**Reply by rijkvanzanten** [reply]
You can totally do this as a plugin. You can create a custom interface that simply renders an iframe with a templated URL, so you can do something like `http://my-staging-url.com/pages/{{ id }}` 🤔 

Certainly interested to have something similar in the platform itself as well 🙂

**Comment by tedsteen** [comment]
The developer experience for this feature in storyblok is pretty nice.
https://www.storyblok.com/tp/add-a-headless-cms-to-svelte-in-5-minutes

Posting for inspiration

**Comment by rijkvanzanten** [comment]
Heya! Thanks for opening this feature request!

This feature request has received over **15 votes** from the community. This means we'll move this feature request to the Under Review state! 

The Core team will schedule a meeting to review this request as soon as possible. The discussion will then be approved or denied. You may or may not be invited to join this meeting with the core team.

_For more information, see our [Feature Request Process](https://docs.directus.io/contributing/feature-request-process.html)._

**Comment by pavelstianko** [comment]
I love this feature! My only concern here is that in the side-by-side layout, you don't see the website in common resolution (it will be like 700x700px) - but new window preview will solve this.

**Reply by rijkvanzanten** [reply]
Yeah! That's why the preview window needs to be resizable / zoomable (so you can set it to something like 1920x1080) and why you can pop it out to a new window :)

**Reply by pavelstianko** [reply]
What about to have some kind of "Do not disturb mode" without other UI elements - to have more space for page preview?

![do-not-disturb](https://github.com/directus/directus/assets/13132639/caa49e5e-4774-48ca-b18c-c63f41c5084d)

**Reply by LasseRosenow** [reply]
@pavelstianko sorry for out of context question, but I noticed in your screenshot the live-preview UI looks soooo much better than in the 10.2 release of directus.

Did you change it yourself, or how did this happen?

**Reply by pavelstianko** [reply]
It is just design idea - not real version :)

**Comment by madc** [comment]
This feature is merged into main now:
https://github.com/directus/directus/pull/17794
